internal/targets/hash_bins.go

```diff
@@ -0,0 +1,86 @@
+package targets
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+const MinDelegationHashPrefixBitLen = 1
+const MaxDelegationHashPrefixBitLen = 32
+
+// hexEncode formats x as a hex string, left padded with zeros to padWidth.
+func hexEncode(x uint64, padWidth int) string {
+	// Benchmarked to be more than 10x faster than padding with Sprintf.
+	s := strconv.FormatUint(x, 16)
+	if len(s) >= padWidth {
+		return s
+	}
+	return strings.Repeat("0", padWidth-len(s)) + s
+}
+
+// HashBin represents a hex prefix range. First should be less than Last.
+type HashBin struct {
+	First uint64
+	Last  uint64
+}
+
+// Name returns the name of the role that signs for the HashBin.
+func (b HashBin) Name(prefix string, padWidth int) string {
+	if b.First == b.Last {
+		return prefix + hexEncode(b.First, padWidth)
+	}
+
+	return prefix + hexEncode(b.First, padWidth) + "-" + hexEncode(b.Last, padWidth)
+}
+
+// Enumerate returns a slice of hash prefixes in the range from First to Last.
+func (b HashBin) Enumerate(padWidth int) []string {
+	n := int(b.Last - b.First + 1)
+	ret := make([]string, int(n))
+
+	x := b.First
+	for i := 0; i < n; i++ {
+		ret[i] = hexEncode(x, padWidth)
+		x++
+	}
+
+	return ret
+}
+
+const bitsPerHexDigit = 4
+
+// HashPrefixLen returns the width of hash prefixes if there are 2^(bitLen)
+// hash bins.
+func HashPrefixLen(bitLen uint8) int {
+	// ceil(bitLen / bitsPerHexDigit)
+	return int((bitLen-1)/bitsPerHexDigit) + 1
```

Comment:
What about the case bitLen = 0? -- you had it before.

Reply:
Is there a use case for using a single bin rather than just a wildcard delegation? Delegations are generally simpler if they can work.

Reply:
I can't think of any use case for a single bin.